[Core] Examiner decoupled from centerIDs

ridz1208 opened the pull request:
This pull request Solves a examiner problem revealed by PR https://github.com/aces/Loris/pull/2644 where examiners completing training will get certified at a random site if they are associated with multiple sites.

This pull request also revamps the logic where the examiner is now certified for instruments regardless of his/her site as certifications should be correlated to examiners rather than examiner/site duo

jstirling91 commented:
Fix PHPCS errors in train module, in both php file and ajax file

jstirling91 commented on SQL/Archive/17.1/2017-03-24-examiner_decoupling_from_site.sql:
```diff
@@ -0,0 +1,25 @@
+
+-- ALTER TABLE certification ADD CONSTRAINT `FK_certifcation_2` FOREIGN KEY (`examinerID`) REFERENCES `examiners` (`examinerID`);
+-- ALTER TABLE examiners ADD CONSTRAINT `U_examiners_1` UNIQUE KEY `full_name` (`full_name`);
+-- ALTER TABLE examiners MODIFY COLUMN `radiologist` tinyint(1) default 0;
```
I don't like the idea of committing out SQL patch code, we should maybe add new infrastructure to better notify the users with this type of patches

AnyhowStep replied:
*Cough* phinx *Cough*

jstirling91 commented:
To discuss on Wednesday, April 5th @11:00 about the implementation. Should we go the route of an examiner is certified for these instruments and all sites they're an examiner at, or should we allow for the ability to certify a subset for instruments in one site and another in a second?

taracampbell commented:
@ridz1208 @jstirling91 Maybe the best route at the moment is to have the tables set up to allow for examiners to have different certifications per site, but in the code for now (i.e. in Training) they will have their certification status updated in the same way across all sites. That way, if we do need the site-specific precision, we can always manually update for specific cases via the backend. And later, front-end control could be added if needed (without having to change the table structure).

ridz1208 commented:
@taracampbell the subject was discussed today in a meeting with @jstirling91 and @driusan. The new changes do not seem to take any functionality away from the current state. Site-specific certification is not juged necessary at the moment so it will be left for a future iteration of improvements. Advantage of this PR are SQL normalization of tables and fix of surrent problem where certifications would be saved semi-randomly for examiners with multiple sites

AnyhowStep commented:
This is one very good example where SQL-only patches fail.

@driusan

AnyhowStep commented:
Null checks should be placed because of the following image.
Could break your patch code.

![image](https://cloud.githubusercontent.com/assets/5655961/25626599/a13f0cea-2f92-11e7-87c5-3a33c7513137.png)

AnyhowStep commented:
Actually, there is a full-SQL solution... If only our database was normalized correctly (3NF, minimally, 6NF for instruments)

ridz1208 commented:
@AnyhowStep is it too late to normalise then implement the full SQL solution ??

AnyhowStep commented:
The normalization would require that the instruments table(s) be normalized to 6NF/EAV. So, yeah, it is too late, lol.

ridz1208 commented:
@AnyhowStep pretty sure 3NF would have been sufficient as we are not tracking history of changes

taracampbell commented:
There is an issue when a user (and therefore associated examiner) does not have a site assigned:

```PHP Warning:  PDOStatement::execute(): SQLSTATE[23000]: Integrity constraint violation: 1048 Column 'centerID' cannot be null in /Users/Tara/Sites/loris/php/libraries/Database.class.inc on line 312
Could not execute INSERT INTO examiners_psc_rel SET `examinerID`=:examinerID,`centerID`=:centerID,`active`=:active,`pending_approval`=:pending_approval. Stack trace#0 /Users/Tara/Sites/loris/php/libraries/Database.class.inc(243): Database->_realinsert('examiners_psc_r...', Array, true)
#1 /Users/Tara/Sites/loris/tools/populate_examiners_psc_rel.php(116): Database->insert('examiners_psc_r...', Array)
#2 {main}
PHP Fatal error:  Uncaught exception 'DatabaseException' with message 'Insert statement did not execute successfully: Array
(
    [0] => 23000
    [1] => 1048
    [2] => Column 'centerID' cannot be null
)
' in /Users/Tara/Sites/loris/php/libraries/Database.class.inc:314
Stack trace:
#0 /Users/Tara/Sites/loris/php/libraries/Database.class.inc(243): Database->_realinsert('examiners_psc_r...', Array, true)
#1 /Users/Tara/Sites/loris/tools/populate_examiners_psc_rel.php(116): Database->insert('examiners_psc_r...', Array)
#2 {main}
  thrown in /Users/Tara/Sites/loris/php/libraries/Database.class.inc on line 314```

taracampbell commented:
Strip examiner names when deleting duplicates - trailing spaces will mess up the script currently.

See Marie Savundranayagam in the CCNA database

MounaSafiHarab commented:
> There is an issue when a user (and therefore associated examiner) does not have a site assigned:
> 
> Could not execute INSERT INTO examiners_psc_rel SET `examinerID`=:examinerID,`centerID`=:centerID,`active`=:active,`pending_approval`=:pending_approval. Stack trace#0 /Users/Tara/Sites/loris/php/libraries/Database.class.inc(243): Database->_realinsert('examiners_psc_r...', Array, true)
> #1 /Users/Tara/Sites/loris/tools/populate_examiners_psc_rel.php(116): Database->insert('examiners_psc_r...', Array)
> #2 {main}
> PHP Fatal error:  Uncaught exception 'DatabaseException' with message 'Insert statement did not execute successfully: Array
> (
>     [0] => 23000
>     [1] => 1048
>     [2] => Column 'centerID' cannot be null
> )
> ' in /Users/Tara/Sites/loris/php/libraries/Database.class.inc:314
> Stack trace:
> #0 /Users/Tara/Sites/loris/php/libraries/Database.class.inc(243): Database->_realinsert('examiners_psc_r...', Array, true)
> #1 /Users/Tara/Sites/loris/tools/populate_examiners_psc_rel.php(116): Database->insert('examiners_psc_r...', Array)
> #2 {main}
>   thrown in /Users/Tara/Sites/loris/php/libraries/Database.class.inc on line 314```

This was reported by @ridz1208 in this redmine task 12673, and @davidblader issued a pull request for this here: #2893

ridz1208 commented:
@taracampbell it finally passed Travis.
Couple new things:
 - examiner table has new userID field. If user is deleted, field is nullified. UserID will be NULL if examiner.full_name does not match exactly user.Real_name. (we can potentially have a small tool which lists all examiners that are not users)
 - I tried uncommenting as many SQL statements as possible. the commented ones need to be run after the php script. I added them as a print out after the php runs successfully.
 - many new null/empty/error checks for the database have been added. I am assuming IBIS and other projects will need them

ridz1208 commented:
@sruthymathew123 It would be great if you can run this as well. 
 - Run sql patch first
 - Run php script `populate_examiner_psc_rel.php`
 - Run remaining Clean-up statements.

sruthymathew123 commented:
@ridz1208 
Ran the patches and ran the php script.
IBIS database throws some errors and I noted it for IBIS fix. Anyway, I managed to run the script on my VM. 
Almost all the duplicates were removed. That is good.
But the remaining Clean up statements are not working as expected.

sruthymathew123 commented:
![screenshot from 2017-07-05 14 45 03](https://user-images.githubusercontent.com/23702452/27879726-89c6541c-6190-11e7-8522-839622304c73.png)

xlecours commented on SQL/Archive/17.1/2017-03-24-examiner_decoupling_from_site.sql:
```diff
@@ -0,0 +1,40 @@
+
+-- ALTER TABLE certification ADD CONSTRAINT `FK_certifcation_2` FOREIGN KEY (`examinerID`) REFERENCES `examiners` (`examinerID`);
+-- ALTER TABLE examiners ADD CONSTRAINT `U_examiners_1` UNIQUE KEY `full_name` (`full_name`);
+
```
This should also remove duplicates in the examiner table. But other table that stores ExaminerID should be update before.

ridz1208 replied:
done by script

xlecours commented on SQL/Archive/17.1/2017-03-24-examiner_decoupling_from_site.sql:
```diff
@@ -0,0 +1,40 @@
+-- NEW COLUMN
+ALTER TABLE examiners ADD COLUMN `userID` int(10) unsigned;
+ALTER TABLE examiners ADD CONSTRAINT `FK_examiners_2` FOREIGN KEY (`userID`) REFERENCES `users` (`ID`) ON DELETE SET NULL ON UPDATE CASCADE;
+UPDATE examiners e SET e.userID=(SELECT ID FROM users WHERE Real_name=e.full_name);
```
This should `TRIM(Real_name)` since e.full_name has been trim

ridz1208 replied:
added the trim however I'm not sure I should actually Update the values in the users table (I will not be adding the SQL `UPDATE users SET Real_name = TRIM(Real_name);` Since I can't predict side effects...

xlecours commented on SQL/Archive/17.1/2017-03-24-examiner_decoupling_from_site.sql:
```diff
@@ -0,0 +1,40 @@
+INSERT examiners_17_0 SELECT * FROM examiners;
+
+-- CLEAN-UP
+UPDATE examiners SET full_name = TRIM(full_name);
```
All instruments Examiner values should be trim also. Right?

xlecours replied:
I have looked at a couple instruments, they seems to hold the ExaminerID. They still need to be updated befor removing duplicate in th eexaminer table.

ridz1208 replied:
do**ne in the php script that omes in this PR

xlecours commented on modules/examiner/php/NDB_Menu_Filter_Form_examiner.class.inc:
```diff
@@ -60,23 +60,25 @@ class NDB_Menu_Filter_Form_Examiner extends NDB_Menu_Filter_Form
-                        LEFT JOIN test_names tn ON (tn.ID = c.testID)
-                        WHERE e.centerID =".$centerID." ";
-
+                        FROM examiners e 
```
The group_by should be done in e.examinerID instead of e.full_name to avoid the infamous 'ONLY_FULL_GROUP_BY' error

ridz1208 replied:
given the new SQL architecture full name should be a unique key and thus obey the functional dependency restrictions of MySQL 5.7... I would need someone to test this however since I'm on 5.7

ridz1208 commented on htdocs/request_account/process_new_account.php:
```diff
@@ -41,8 +41,8 @@
 
-$res = array();
-$DB->select("SELECT Name, CenterID FROM psc", $res);
+$res       = array();
```
I will hold off on this for this PR

I did add the comment to https://github.com/aces/Loris/pull/2961 which is cleaning up this particular class

ridz1208 commented on htdocs/request_account/process_new_account.php:
```diff
@@ -212,6 +196,48 @@
                 )
             );
         }
+
```
yes

ridz1208 commented:
@xlecours @driusan @jstirling91 

new changes done should be covering all possible cases at this point.

@sruthymathew123 
ready for testing as soon as I rebase it.
 - added handling for lowercase/uppercase
 - added handling for trailing spaces
- added handling for multi-spaces between firs last names

sruthymathew123 commented:
@ridz1208 Tested on IBIS data set and found some cleanup sql patches didn't run successfully. Attaching those error log for your perusal. 

1.  ALTER TABLE examiners DROP FOREIGN KEY FK_examiners_1	
Error Code: 1025. Error on rename of './smathew_ibis/examiners' to './smathew_ibis/#sql2-7517-5dcb4' (errno: 152)

2. ALTER TABLE examiners DROP KEY FK_examiners_1	
Error Code: 1091. Can't DROP 'FK_examiners_1'; check that column/key exists

3.   ALTER TABLE certification ADD CONSTRAINT `FK_certifcation_2` FOREIGN KEY (`examinerID`) REFERENCES `examiners` (`examinerID`)	
Error Code: 1005. Can't create table 'smathew_ibis.#sql-7517_5dcb4' (errno: 121)

NOTE: Other things like the SQL patches and the script ran successfully. The script also identified examiner vulnerabilities in some instruments which the dataset should fix.

sruthymathew123 commented:
@ridz1208 It seems that the problem lies with foreign key constraints.

xlecours commented on modules/examiner/php/NDB_Menu_Filter_Form_examiner.class.inc:
```diff
@@ -258,6 +300,28 @@ class NDB_Menu_Filter_Form_Examiner extends NDB_Menu_Filter_Form
+     *
+     * @return string   ID of the examiner
+     */
+    function getExaminerID($fullName)
```
This should be in the Examiner class.

ridz1208 replied:
https://redmine.cbrain.mcgill.ca/issues/12875

xlecours commented on php/libraries/Database.class.inc:
```diff
@@ -285,18 +302,25 @@ class Database
-        $query  = "INSERT INTO $table SET ";
+        if ($ignore) {
+            $query = "INSERT IGNORE INTO $table SET ";
+            $prepQ = "INSERT IGNORE INTO $table SET ";
```
$prepQ = $query after this if block?

ridz1208 replied:
@xlecours just didn't wanna change too much stuff

ridz1208 commented:
@sruthymathew123 @gluneau IBIS does not seem to have the correct Foreign Keys on the original table... It seems like a patch is missing, however, since the `FK_examiners_1` is in the schema I have to add the drop statements here !! keep note of this when releasing 17.1 on IBIS

sruthymathew123 commented:
@gluneau @ridz1208 The only problem that I noticed on testing this PR is that the cleanup SQL statements didn't work well on  IBIS examiners schema as it is missing foreign key constraint.

Apart from that, all seems good for me. So as this not any blocking issue, I am giving passed manual tests label.

ridz1208 commented:
@driusan this was approved by @xlecours but review was dismissed automatically after rebase

ridz1208 commented:
@xlecours please don't hate me

driusan commented on SQL/Archive/17.1/2017-03-24-examiner_decoupling_from_site.sql:
```diff
@@ -0,0 +1,43 @@
+DROP TABLE IF EXISTS `examiners_psc_rel`;
```
I don't think patches that are intended to run on production should have `DROP TABLE IF EXISTS`.. it's a recipe for accidental data loss.

ridz1208 replied:
I live for accidental data loss

driusan commented on htdocs/request_account/process_new_account.php:
```diff
@@ -212,6 +196,48 @@
         }
+
+        if ($_REQUEST['examiner']=='on') {
+            $rad =0;
```
has this been run through phpcs? This formatting doesn't look right.

ridz1208 replied:
its passing so... yes ?

ridz1208 replied:
will run davecbf

driusan commented on modules/examiner/php/NDB_Form_examiner.class.inc:
```diff
@@ -54,16 +54,23 @@ class NDB_Form_Examiner extends NDB_Form
+                $centerIDs[] = $row['centerID'];
+            }
 
+            // Access is only granted if user ENGLOBES all sites of examiner
```
Why is "ENGLOBES" all caps? And is there a more.. well known.. word that can be used instead so that people don't need a thesaurus to understand this comment?

ridz1208 replied:
... -_-

driusan commented on modules/examiner/php/NDB_Form_examiner.class.inc:
```diff
@@ -54,16 +54,23 @@ class NDB_Form_Examiner extends NDB_Form
             return $user->hasPermission('examiner_view')
                 && (($user->hasPermission('examiner_multisite')
-                || in_array($centerID, $user->getData('CenterIDs'))));
+                || empty(array_diff($centerIDs, $user->getData('CenterIDs')))));
```
This line is starting to get convoluted.. maybe add a `$user->hasAllCenterIDs(array())` (or `$user->englobesCenterIDs()`) function? It seems like something that will be used in multiple places and should live in the user class.

ridz1208 replied:
https://redmine.cbrain.mcgill.ca/issues/12948

ridz1208 commented:
@driusan all changes done

driusan commented:
@ridz1208 This has conflicts and needs to be rebased.

driusan commented on tools/populate_examiners_psc_rel.php:
```diff
@@ -0,0 +1,213 @@
+$pscCheck = $db->pselect(
+    "SELECT * FROM examiners WHERE centerID NOT IN (SELECT centerID FROM psc)",
+    array()
+);
```
this sounds like there needs to be a foreign key added..

driusan commented on tools/populate_examiners_psc_rel.php:
```diff
@@ -0,0 +1,213 @@
+
+print_r($examinersOrganized);
+
+echo "\nA copy of the `examiners` table has been created as `examiners_17.0`. \n\n";
```
I might be not seeing it above, but this statement sounds like a lie. Where does this happen?

ridz1208 replied:
This is just reassuring the user that the data has been copied (in the SQL patch)

driusan commented on tools/populate_examiners_psc_rel.php:
```diff
@@ -0,0 +1,213 @@
+-- ALTER TABLE examiners DROP KEY FK_examiners_1;\n
+-- ALTER TABLE examiners DROP COLUMN `CenterID`;\n
+-- ALTER TABLE examiners DROP COLUMN `active`;\n
+-- ALTER TABLE examiners DROP COLUMN `pending_approval`;\n
```
If these are just being printed and not executed, why do they start with an SQL comment delimiter? (`--`)? It seems confusing, and people will probably copy/paste it and not realize it doesn't do anything..

MounaSafiHarab commented on SQL/Archive/17.1/2017-03-24-examiner_decoupling_from_site.sql:
```diff
@@ -0,0 +1,42 @@
+ALTER TABLE examiners ADD COLUMN `userID` int(10) unsigned;
+ALTER TABLE examiners ADD CONSTRAINT `FK_examiners_2` FOREIGN KEY (`userID`) REFERENCES `users` (`ID`) ON DELETE SET NULL ON UPDATE CASCADE;
+
+-- ** WARNING: The next lines will DROP column CenterID from the users table **
```
@ridz1208 
copy and paste is the culprit in this Warning message (I just saw this as I am sourcing your patch on my VM).
